### DISCUSSION TITLE
Managing TaskCancellationTokens in a central service in ASP.NET

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#116] [General]
# Managing TaskCancellationTokens in a central service in ASP.NET

In async code it is very common to pass a CancellationToken to the method that is being awaited. This allows the caller to cancel the operation if it is no longer needed. But this will lead to some ceremonial code that is repeated in every method. In this article I will show you how to manage CancellationToken in a central service in ASP.NET.

https://steven-giesel.com/blogPost/cacc60e5-4fbf-473a-a5af-c9fb333d9183/managing-taskcancellationtokens-in-a-central-service-in-aspnet

**Comment by toupswork** [comment]
I've thought about this before. Back in the day, we could leverage thread local storage to create a class like CancellationTokenContext, and then whichever method needs it, we'd call a static function on that class to get the cancellation token. We don't do that anymore because it's not testable, and the new way is to use AsynLocal or ThreadStatic. 

That's how IHttpContextAccessor works. I'm thinking you could create a dependency like that which stores the cancellation token and makes it available whenever you need it later rather than passing it down to each method. 

It's not a big issue for me, because there aren't any hops (or rarely very few) where I have to pass the cancellationtoken to methods that don't themselves use it.

**Reply by toupswork** [reply]
```C#
public interface ICancellationTokenAccessor
{
    CancellationToken? Token { get; set; }
}
```
```C#
public class CancellationTokenAccessor : ICancellationTokenAccessor
{
    private AsyncLocal<CancellationToken?> _cancellationToken = new AsyncLocal<CancellationToken?>();

    public CancellationToken? Token
    {
        get => _cancellationToken.Value;
        set => _cancellationToken.Value = value;
    }
}
```

```C#
public class CancellationTokenMiddleware
{
    private readonly RequestDelegate _next;
    private readonly ICancellationTokenAccessor _tokenAccessor;

    public CancellationTokenMiddleware(RequestDelegate next, ICancellationTokenAccessor tokenAccessor)
    {
        _next = next;
        _tokenAccessor = tokenAccessor;
    }

    public async Task InvokeAsync(HttpContext context)
    {
        // Set the cancellation token from the current HTTP context
        _tokenAccessor.Token = context.RequestAborted;

        try
        {
            await _next(context);
        }
        finally
        {
            // Clear the token after the request is complete to avoid potential leaks
            _tokenAccessor.Token = null;
        }
    }
}
```

```C#
public void ConfigureServices(IServiceCollection services)
{
    services.AddSingleton<ICancellationTokenAccessor, CancellationTokenAccessor>();
}

public void Configure(IApplicationBuilder app)
{
    app.UseMiddleware<CancellationTokenMiddleware>();
}
```

```C#
public class MyService
{
    private readonly ICancellationTokenAccessor _tokenAccessor;

    public MyService(ICancellationTokenAccessor tokenAccessor)
    {
        _tokenAccessor = tokenAccessor;
    }

    public async Task SomeLongRunningOperation()
    {
        // Access the cancellation token
        var token = _tokenAccessor.Token ?? CancellationToken.None;

        // Use the token in your async operations
        await Task.Delay(5000, token);
    }
}
```

**Reply by toupswork** [reply]
The scoped lifetime and the absence of asynclocal might pose issues in the article sample

**Reply by linkdotnet** [reply]
Interesting! Didn't even think of stuff like `AsyncLocal`. The question is what does `AsyncLocal` offer over "just" passing the token to a container which then can be used.
The consuming parts are readonly (in terms of the token at least) therefore: Is there a need for `AsyncLocal` in the first place?

**Reply by toupswork** [reply]
It can cause it to be overwritten if another middleware runs after the cancellation token is grabbed. 
You probably already know that TLS is often used to prevent passing parameters from method to method. So, asynclocal and threadstatic are how TLS is done now.  IHttpContextAccessor uses AsyncLocal

**Reply by toupswork** [reply]
If you need help on what to write about next, I have an idea. How about writing custom Linq extension methods that allow you to pass in a context without creating a closure; e.g., 

`.WhereWithContext<TContext,TSource>(this TSource @this, TContext context, Func<TContext, TSource, bool> predicate);`

I just learned something new the other day: even when you put `static` in front of your lambda, that doesn't have any effect on the compiler. It's simply a pre-compilation keyword that helps you identify any closures by displaying a syntax error.

**Reply by linkdotnet** [reply]
That is an interesting idea and the dotnet used that a lot (for example string.Create and friends)!
Noted!

**Reply by toupswork** [reply]
Wait, sorry... one more thing I wanted to ask you about. Did you already know about `RuntimeHelpers.GetUninitializedObject()`, where it lets you bypass the constructor? It's super helpful in unit tests when combined with `UnsafeAccessor` for helping to reduce the amount of fake data you have to create.

So here I have a `record` type that is nested and complex, so painful to instantiate, but for my test, I know there are only a few fields that it needs, so I use the `with` keyword to fill those in
![image](https://github.com/user-attachments/assets/d44b1494-bde5-4e0b-a8ff-dafd331b2035)

In this example, I set my mock to return a collection, but I return the ref to that item in that collection so my test can take the pointer to a pointer and re-point the array pointer to another object
![image](https://github.com/user-attachments/assets/69321cf2-dcb6-43ec-8519-bb9f755fbe5b)

Then with some marshalling classes or the UnsafeAccessor I can mutate immutable record types, or I can use marshalling classes to overwrite the memory address of an instance of an object on the heap—note, the object instance itself and not a pointer to it.. 
![image](https://github.com/user-attachments/assets/5303d5cd-0d46-4c93-bd86-fd7dcc4f2d37)

**Reply by linkdotnet** [reply]
Interesting approach! And no - never came across that method. Maybe even something for high-performance scenarios.
It kinda behaves like `new MyStruct` vs `default(MyStruct)` where the latter one never goes through the c'tor of the struct!

I try to avoid reflection-ish approaches as much as possible and for tests I normally resort to a Builder that creates the object in question.

**Reply by linkdotnet** [reply]
Did you find any other use case? In my head I can only come up with doing some trickery because of bad design decision.

imsgine a service with many many ctor parameters or ctor logic that is quite expensive .
With this approach you go around.

**Reply by toupswork** [reply]
Just to be clear, I mostly use this approach to expedite writing unit tests where it would take a lot of time to create instances of models with dummy data, because some models have a large graph of fields and nested objects. So, I wouldn't use this for a service.  But outside of unit tests, I will sometimes create a static readonly property of an empty record instance using that RuntimeHelper, for those occasions when I need to return a non-null instance of it, or to use that empty instance and apply the `with` operator to add values to some of the fields.

**Reply by toupswork** [reply]
I remembered something. This team at my company wrote this crappy security library that I had to use and test. They have a User class you must use, and they put in the constructor a call to the asp.net runtime to get the current user's JWT. And of course it blows up in unit tests. So this helped tremendously. Using Reflection or Activator still requires invoking a ctor last time I checked. And there was no other ctor.

**Reply by linkdotnet** [reply]
Wow - yeah well. I guess using the `User` object in tests can't then really be done (given that dependency and design of the object - the ctor isn't the only issue)?
